// kubeflow/esm2015/lib/conditions-table/conditions-table.component.js
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { STATUS_TYPE } from '../resource-table/status/types';
import { generateConfig } from './config';
import * as i0 from "@angular/core";
import * as i1 from "../resource-table/resource-table.component";
export class ConditionsTableComponent {
    constructor() {
        this.conditionsPrv = [];
        this.config = generateConfig();
    }
    set title(t) {
        this.config.title = t;
    }
    set conditions(cs) {
        this.conditionsPrv = JSON.parse(JSON.stringify(cs));
        // parse the status. It should be ready only if it was True
        for (const condition of this.conditionsPrv) {
            condition.statusPhase = STATUS_TYPE.WARNING;
            if (condition.status === 'True') {
                condition.statusPhase = STATUS_TYPE.READY;
            }
            condition.statusMessage = condition.status;
        }
    }
    get conditions() {
        return this.conditionsPrv;
    }
    conditionsTrackByFn(index, c) {
        return `${c.type}/${c.lastTransitionTime}`;
    }
}
ConditionsTableComponent.ɵfac = function ConditionsTableComponent_Factory(t) { return new (t || ConditionsTableComponent)(); };
ConditionsTableComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: ConditionsTableComponent, selectors: [["lib-conditions-table"]], inputs: { title: "title", conditions: "conditions" }, decls: 1, vars: 3, consts: [[3, "config", "data", "trackByFn"]], template: function ConditionsTableComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelement(0, "lib-resource-table", 0);
    } if (rf & 2) {
        i0.ɵɵproperty("config", ctx.config)("data", ctx.conditions)("trackByFn", ctx.conditionsTrackByFn);
    } }, directives: [i1.ResourceTableComponent], encapsulation: 2, changeDetection: 0 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ConditionsTableComponent, [{
        type: Component,
        args: [{
                selector: 'lib-conditions-table',
                templateUrl: './conditions-table.component.html',
                changeDetection: ChangeDetectionStrategy.OnPush,
                styleUrls: [],
            }]
    }], null, { title: [{
            type: Input
        }], conditions: [{
            type: Input
        }] }); })();
