// kubeflow/esm2015/lib/resource-table/resource-table.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ResourceTableComponent } from './resource-table.component';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule } from '@angular/material/icon';
import { MatMenuModule } from '@angular/material/menu';
import { MatTableModule } from '@angular/material/table';
import { StatusComponent } from './status/status.component';
import { MatTooltipModule } from '@angular/material/tooltip';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { ActionComponent } from './action/action.component';
import { MatChipsModule } from '@angular/material/chips';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { ActionButtonComponent } from './action-button/action-button.component';
import { IconModule } from '../icon/icon.module';
import { TableComponent } from './table/table.component';
import { DateTimeModule } from '../date-time/date-time.module';
import { PopoverModule } from '../popover/popover.module';
import { TableChipsListComponent } from './chips-list/chips-list.component';
import { ComponentValueComponent } from './component-value/component-value.component';
import { PortalModule } from '@angular/cdk/portal';
import * as i0 from "@angular/core";
export class ResourceTableModule {
}
ResourceTableModule.ɵfac = function ResourceTableModule_Factory(t) { return new (t || ResourceTableModule)(); };
ResourceTableModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: ResourceTableModule });
ResourceTableModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            MatTableModule,
            MatTooltipModule,
            MatProgressSpinnerModule,
            MatDividerModule,
            MatCardModule,
            MatButtonModule,
            MatChipsModule,
            MatMenuModule,
            PortalModule,
            FontAwesomeModule,
            MatIconModule,
            IconModule,
            DateTimeModule,
            PopoverModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ResourceTableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MatTableModule,
                    MatTooltipModule,
                    MatProgressSpinnerModule,
                    MatDividerModule,
                    MatCardModule,
                    MatButtonModule,
                    MatChipsModule,
                    MatMenuModule,
                    PortalModule,
                    FontAwesomeModule,
                    MatIconModule,
                    IconModule,
                    DateTimeModule,
                    PopoverModule,
                ],
                declarations: [
                    ResourceTableComponent,
                    StatusComponent,
                    ActionComponent,
                    ActionButtonComponent,
                    TableChipsListComponent,
                    TableComponent,
                    ComponentValueComponent,
                ],
                exports: [ResourceTableComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(ResourceTableModule, { declarations: [ResourceTableComponent,
        StatusComponent,
        ActionComponent,
        ActionButtonComponent,
        TableChipsListComponent,
        TableComponent,
        ComponentValueComponent], imports: [CommonModule,
        MatTableModule,
        MatTooltipModule,
        MatProgressSpinnerModule,
        MatDividerModule,
        MatCardModule,
        MatButtonModule,
        MatChipsModule,
        MatMenuModule,
        PortalModule,
        FontAwesomeModule,
        MatIconModule,
        IconModule,
        DateTimeModule,
        PopoverModule], exports: [ResourceTableComponent] }); })();
