// kubeflow/esm2015/lib/resource-table/resource-table.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { ActionEvent, TABLE_THEME, } from './types';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/card";
import * as i3 from "@angular/material/divider";
import * as i4 from "./table/table.component";
import * as i5 from "@angular/material/button";
import * as i6 from "@angular/material/icon";
function ResourceTableComponent_mat_card_1_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵlistener("click", function ResourceTableComponent_mat_card_1_button_4_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(2); return ctx_r3.newButtonTriggered(); });
    i0.ɵɵelementStart(1, "mat-icon", 8);
    i0.ɵɵtext(2, "add");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate1(" ", ctx_r2.config == null ? null : ctx_r2.config.newButtonText, " ");
} }
function ResourceTableComponent_mat_card_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "mat-card", 3);
    i0.ɵɵelementStart(1, "div", 4);
    i0.ɵɵelementStart(2, "div");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(4, ResourceTableComponent_mat_card_1_button_4_Template, 4, 1, "button", 5);
    i0.ɵɵelementEnd();
    i0.ɵɵelement(5, "mat-divider");
    i0.ɵɵelementStart(6, "mat-card-content");
    i0.ɵɵelement(7, "lib-table", 6);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("width", ctx_r0.totalWidth)("min-width", ctx_r0.minTableWidth);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.config == null ? null : ctx_r0.config.title);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r0.config == null ? null : ctx_r0.config.newButtonText);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("config", ctx_r0.config)("data", ctx_r0.data)("trackByFn", ctx_r0.trackByFn)("emitter", ctx_r0.actionsEmitter);
} }
function ResourceTableComponent_div_2_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵlistener("click", function ResourceTableComponent_div_2_button_4_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(2); return ctx_r6.newButtonTriggered(); });
    i0.ɵɵelementStart(1, "mat-icon", 8);
    i0.ɵɵtext(2, "add");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate1(" ", ctx_r5.config == null ? null : ctx_r5.config.newButtonText, " ");
} }
function ResourceTableComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵelementStart(1, "div", 9);
    i0.ɵɵelementStart(2, "p");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(4, ResourceTableComponent_div_2_button_4_Template, 4, 1, "button", 5);
    i0.ɵɵelementEnd();
    i0.ɵɵelement(5, "mat-divider");
    i0.ɵɵelement(6, "lib-table", 6);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("width", ctx_r1.totalWidth)("min-width", ctx_r1.minTableWidth);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.config == null ? null : ctx_r1.config.title);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.config == null ? null : ctx_r1.config.newButtonText);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("config", ctx_r1.config)("data", ctx_r1.data)("trackByFn", ctx_r1.trackByFn)("emitter", ctx_r1.actionsEmitter);
} }
export class ResourceTableComponent {
    constructor() {
        // Whenever a button in a row is pressed the component will emit an event
        // with information regarding the button that was pressed as well as the
        // row's object.
        this.actionsEmitter = new EventEmitter();
        this.TABLE_THEME = TABLE_THEME;
    }
    ngOnInit() { }
    actionTriggered(e) {
        // Forward the emitted ActionEvent
        this.actionsEmitter.emit(e);
    }
    newButtonTriggered() {
        const ev = new ActionEvent('newResourceButton', {});
        this.actionsEmitter.emit(ev);
    }
    linkClicked(field, data) {
        const ev = new ActionEvent(`${field}:link`, data);
        this.actionsEmitter.emit(ev);
    }
    get minTableWidth() {
        // Review: This will break if the config is an other falsy value
        // https://developer.mozilla.org/en-US/docs/Glossary/Falsy
        if (typeof this.config === 'undefined') {
            return '600px';
        }
        return `${this.config.columns.length * 100}px`;
    }
    get totalWidth() {
        if (!this.config || !this.config.width) {
            return 'fit-content';
        }
        return this.config.width;
    }
    get tableTheme() {
        if (!this.config || !this.config.theme) {
            return TABLE_THEME.CARD;
        }
        return this.config.theme;
    }
}
ResourceTableComponent.ɵfac = function ResourceTableComponent_Factory(t) { return new (t || ResourceTableComponent)(); };
ResourceTableComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: ResourceTableComponent, selectors: [["lib-resource-table"]], inputs: { config: "config", data: "data", trackByFn: "trackByFn" }, outputs: { actionsEmitter: "actionsEmitter" }, decls: 3, vars: 4, consts: [[1, "center-flex"], ["class", "mat-elevation-z4", 3, "width", "min-width", 4, "ngIf"], [3, "width", "min-width", 4, "ngIf"], [1, "mat-elevation-z4"], [1, "header", "card-title-padding"], ["mat-button", "", "id", "newResource", "color", "primary", "class", "right", 3, "click", 4, "ngIf"], [3, "config", "data", "trackByFn", "emitter"], ["mat-button", "", "id", "newResource", "color", "primary", 1, "right", 3, "click"], [1, "new-resource-button"], [1, "header"]], template: function ResourceTableComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, ResourceTableComponent_mat_card_1_Template, 8, 10, "mat-card", 1);
        i0.ɵɵtemplate(2, ResourceTableComponent_div_2_Template, 7, 10, "div", 2);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵclassProp("space-top", ctx.tableTheme == ctx.TABLE_THEME.CARD);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.tableTheme == ctx.TABLE_THEME.CARD);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.tableTheme == ctx.TABLE_THEME.FLAT);
    } }, directives: [i1.NgIf, i2.MatCard, i3.MatDivider, i2.MatCardContent, i4.TableComponent, i5.MatButton, i6.MatIcon], styles: [".right[_ngcontent-%COMP%]{margin-left:auto}.space-top[_ngcontent-%COMP%]{padding-top:1.5rem}mat-card[_ngcontent-%COMP%]{padding:0;margin:0 0 50px}mat-toolbar[_ngcontent-%COMP%]{background:#fff}.header[_ngcontent-%COMP%]{display:flex;align-items:center;height:64px;font-weight:400;font-size:20px}.card-title-padding[_ngcontent-%COMP%]{padding:0 16px}.mat-icon[_ngcontent-%COMP%]{line-height:.85}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ResourceTableComponent, [{
        type: Component,
        args: [{
                selector: 'lib-resource-table',
                templateUrl: './resource-table.component.html',
                styleUrls: ['./resource-table.component.scss'],
            }]
    }], function () { return []; }, { config: [{
            type: Input
        }], data: [{
            type: Input
        }], trackByFn: [{
            type: Input
        }], actionsEmitter: [{
            type: Output
        }] }); })();
