// kubeflow/esm2015/lib/services/backend/backend.service.js
import { Injectable } from '@angular/core';
import { HttpErrorResponse, } from '@angular/common/http';
import { throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { SnackType } from '../../snack-bar/types';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../../snack-bar/snack-bar.service";
export class BackendService {
    constructor(http, snackBar) {
        this.http = http;
        this.snackBar = snackBar;
        this.apiUrl = '';
    }
    // GETers
    getUsername() {
        const url = `info`;
        return this.http.get(url).pipe(catchError(error => this.handleError(error, false)), map((data) => data.user));
    }
    getNamespaces(showSnackBar = true, url) {
        return this.http.get(url ? url : 'api/namespaces').pipe(catchError(error => this.handleError(error, showSnackBar)), map((data) => data.namespaces
            ? data.namespaces
            : data));
    }
    getStorageClasses(showSnackBar = true) {
        // Get existing PVCs in a namespace
        const url = `api/storageclasses`;
        return this.http.get(url).pipe(catchError(error => this.handleError(error, showSnackBar)), map((data) => {
            return data.storageClasses;
        }));
    }
    getDefaultStorageClass(showSnackBar = true) {
        const url = `api/storageclasses/default`;
        return this.http.get(url).pipe(catchError(error => this.handleError(error, showSnackBar)), map((data) => {
            return data.defaultStorageClass;
        }));
    }
    // ---------------------------Error Handling---------------------------------
    getBackendErrorLog(error) {
        if (error.error === null) {
            return error.message;
        }
        else {
            // Show the message the backend has sent
            return error.error.log;
        }
    }
    getErrorMessage(error) {
        if (typeof error === 'string') {
            return error;
        }
        if (error instanceof HttpErrorResponse) {
            if (this.getBackendErrorLog(error) !== undefined) {
                return this.getBackendErrorLog(error);
            }
            return `${error.status}: ${error.message}`;
        }
        if (error instanceof ErrorEvent) {
            return error.message;
        }
        return `Unexpected error encountered`;
    }
    getSnackErrorMessage(error) {
        if (typeof error === 'string') {
            return $localize `An error occurred: ${error}`;
        }
        if (error.error instanceof ErrorEvent) {
            return $localize `Client error: ${error.error.message}`;
        }
        if (error instanceof HttpErrorResponse) {
            // In case of status code 0 or negative, Http module couldn't
            // connect to the backend
            if (error.status <= 0) {
                return $localize `Could not connect to the backend.`;
            }
            return `[${error.status}] ${this.getBackendErrorLog(error)}\n${error.url}`;
        }
        if (error instanceof ErrorEvent) {
            return error.message;
        }
        return $localize `Unexpected error encountered`;
    }
    handleError(error, showSnackBar = true) {
        // The backend returned an unsuccessful response code.
        // The response body may contain clues as to what went wrong,
        console.error(error);
        if (showSnackBar) {
            this.snackBar.open(this.getSnackErrorMessage(error), SnackType.Error);
        }
        return throwError(this.getErrorMessage(error));
    }
}
BackendService.ɵfac = function BackendService_Factory(t) { return new (t || BackendService)(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.SnackBarService)); };
BackendService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: BackendService, factory: BackendService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(BackendService, [{
        type: Injectable,
        args: [{
                providedIn: 'root',
            }]
    }], function () { return [{ type: i1.HttpClient }, { type: i2.SnackBarService }]; }, null); })();
