// @angular/cdk-experimental/esm2015/combobox/combobox-popup.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, ElementRef, Inject, InjectionToken, Input, Optional } from '@angular/core';
import { CdkComboboxPanel } from './combobox-panel';
export const PANEL = new InjectionToken('CdkComboboxPanel');
let nextId = 0;
export class CdkComboboxPopup {
    constructor(_elementRef, _parentPanel) {
        this._elementRef = _elementRef;
        this._parentPanel = _parentPanel;
        this._role = 'dialog';
        this.id = `cdk-combobox-popup-${nextId++}`;
    }
    get role() {
        return this._role;
    }
    set role(value) {
        this._role = value;
    }
    get firstFocus() {
        return this._firstFocusElement;
    }
    set firstFocus(id) {
        this._firstFocusElement = id;
    }
    ngOnInit() {
        this.registerWithPanel();
    }
    registerWithPanel() {
        if (this._parentPanel === null || this._parentPanel === undefined) {
            this._explicitPanel._registerContent(this.id, this._role);
        }
        else {
            this._parentPanel._registerContent(this.id, this._role);
        }
    }
    focusFirstElement() {
        if (this._firstFocusElement) {
            this._firstFocusElement.focus();
        }
        else {
            this._elementRef.nativeElement.focus();
        }
    }
}
CdkComboboxPopup.decorators = [
    { type: Directive, args: [{
                selector: '[cdkComboboxPopup]',
                exportAs: 'cdkComboboxPopup',
                host: {
                    'class': 'cdk-combobox-popup',
                    '[attr.role]': 'role',
                    '[id]': 'id',
                    'tabindex': '-1',
                    '(focus)': 'focusFirstElement()'
                }
            },] }
];
CdkComboboxPopup.ctorParameters = () => [
    { type: ElementRef },
    { type: CdkComboboxPanel, decorators: [{ type: Optional }, { type: Inject, args: [PANEL,] }] }
];
CdkComboboxPopup.propDecorators = {
    role: [{ type: Input }],
    firstFocus: [{ type: Input }],
    id: [{ type: Input }],
    _explicitPanel: [{ type: Input, args: ['parentPanel',] }]
};
