// kubeflow/esm2015/lib/details-list/details-list.component.js
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/divider";
import * as i3 from "./details-list-item/details-list-item.component";
function DetailsListComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 4);
    i0.ɵɵelementStart(1, "p");
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r0.title);
} }
function DetailsListComponent_mat_divider_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "mat-divider");
} }
function DetailsListComponent_div_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelement(1, "lib-details-list-item", 6);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const entry_r3 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("key", entry_r3.key)("value", entry_r3.value)("keyTooltip", entry_r3.keyTooltip)("valueTooltip", entry_r3.valueTooltip)("icon", entry_r3.icon)("valueType", entry_r3.valueType)("chipsList", entry_r3.chipsList);
} }
function DetailsListComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 5);
    i0.ɵɵtemplate(1, DetailsListComponent_div_3_ng_container_1_Template, 2, 7, "ng-container", 2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const entry_r3 = ctx.$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", entry_r3);
} }
export class DetailsListComponent {
    constructor() {
        this.entries = [];
        this.topDivider = true;
    }
}
DetailsListComponent.ɵfac = function DetailsListComponent_Factory(t) { return new (t || DetailsListComponent)(); };
DetailsListComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: DetailsListComponent, selectors: [["lib-details-list"]], inputs: { entries: "entries", topDivider: "topDivider", title: "title" }, decls: 4, vars: 3, consts: [[1, "list"], ["class", "list-header", 4, "ngIf"], [4, "ngIf"], ["class", "list-entry", 4, "ngFor", "ngForOf"], [1, "list-header"], [1, "list-entry"], [3, "key", "value", "keyTooltip", "valueTooltip", "icon", "valueType", "chipsList"]], template: function DetailsListComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtemplate(1, DetailsListComponent_div_1_Template, 3, 1, "div", 1);
        i0.ɵɵtemplate(2, DetailsListComponent_mat_divider_2_Template, 1, 0, "mat-divider", 2);
        i0.ɵɵtemplate(3, DetailsListComponent_div_3_Template, 2, 1, "div", 3);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.title);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.topDivider);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngForOf", ctx.entries);
    } }, directives: [i1.NgIf, i1.NgForOf, i2.MatDivider, i3.DetailsListItemComponent], styles: [".list[_ngcontent-%COMP%]{display:inline-block;width:100%}.list-header[_ngcontent-%COMP%]{display:flex;align-items:center;height:64px;font-weight:400;font-size:20px}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DetailsListComponent, [{
        type: Component,
        args: [{
                selector: 'lib-details-list',
                templateUrl: './details-list.component.html',
                styleUrls: ['./details-list.component.scss'],
            }]
    }], null, { entries: [{
            type: Input
        }], topDivider: [{
            type: Input
        }], title: [{
            type: Input
        }] }); })();
