// kubeflow/esm2015/lib/heading-subheading-row/heading-subheading-row.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HeadingSubheadingRowComponent } from './heading-subheading-row.component';
import { MatTooltipModule } from '@angular/material/tooltip';
import * as i0 from "@angular/core";
export class HeadingSubheadingRowModule {
}
HeadingSubheadingRowModule.ɵfac = function HeadingSubheadingRowModule_Factory(t) { return new (t || HeadingSubheadingRowModule)(); };
HeadingSubheadingRowModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: HeadingSubheadingRowModule });
HeadingSubheadingRowModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, MatTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(HeadingSubheadingRowModule, [{
        type: NgModule,
        args: [{
                declarations: [HeadingSubheadingRowComponent],
                imports: [CommonModule, MatTooltipModule],
                exports: [HeadingSubheadingRowComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(HeadingSubheadingRowModule, { declarations: [HeadingSubheadingRowComponent], imports: [CommonModule, MatTooltipModule], exports: [HeadingSubheadingRowComponent] }); })();
