// @angular/cdk-experimental/esm2015/column-resize/column-resize-module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CdkColumnResize } from './column-resize-directives/column-resize';
import { CdkColumnResizeFlex } from './column-resize-directives/column-resize-flex';
import { CdkDefaultEnabledColumnResize } from './column-resize-directives/default-enabled-column-resize';
import { CdkDefaultEnabledColumnResizeFlex } from './column-resize-directives/default-enabled-column-resize-flex';
/**
 * One of two NgModules for use with CdkColumnResize.
 * When using this module, columns are resizable by default.
 */
import * as ɵngcc0 from '@angular/core';
export class CdkColumnResizeDefaultEnabledModule {
}
CdkColumnResizeDefaultEnabledModule.ɵfac = function CdkColumnResizeDefaultEnabledModule_Factory(t) { return new (t || CdkColumnResizeDefaultEnabledModule)(); };
CdkColumnResizeDefaultEnabledModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CdkColumnResizeDefaultEnabledModule });
CdkColumnResizeDefaultEnabledModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkColumnResizeDefaultEnabledModule, [{
        type: NgModule,
        args: [{
                declarations: [CdkDefaultEnabledColumnResize, CdkDefaultEnabledColumnResizeFlex],
                exports: [CdkDefaultEnabledColumnResize, CdkDefaultEnabledColumnResizeFlex]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkColumnResizeDefaultEnabledModule, { declarations: function () { return [CdkDefaultEnabledColumnResize, CdkDefaultEnabledColumnResizeFlex]; }, exports: function () { return [CdkDefaultEnabledColumnResize, CdkDefaultEnabledColumnResizeFlex]; } }); })();
/**
 * One of two NgModules for use with CdkColumnResize.
 * When using this module, columns are not resizable by default.
 */
export class CdkColumnResizeModule {
}
CdkColumnResizeModule.ɵfac = function CdkColumnResizeModule_Factory(t) { return new (t || CdkColumnResizeModule)(); };
CdkColumnResizeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CdkColumnResizeModule });
CdkColumnResizeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkColumnResizeModule, [{
        type: NgModule,
        args: [{
                declarations: [CdkColumnResize, CdkColumnResizeFlex],
                exports: [CdkColumnResize, CdkColumnResizeFlex]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkColumnResizeModule, { declarations: function () { return [CdkColumnResize, CdkColumnResizeFlex]; }, exports: function () { return [CdkColumnResize, CdkColumnResizeFlex]; } }); })();
