// @angular/cdk-experimental/esm2015/column-resize/column-resize-notifier.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Originating source of column resize events within a table.
 * @docs-private
 */
import * as ɵngcc0 from '@angular/core';
export class ColumnResizeNotifierSource {
    constructor() {
        /** Emits when an in-progress resize is canceled. */
        this.resizeCanceled = new Subject();
        /** Emits when a resize is applied. */
        this.resizeCompleted = new Subject();
        /** Triggers a resize action. */
        this.triggerResize = new Subject();
    }
}
ColumnResizeNotifierSource.ɵfac = function ColumnResizeNotifierSource_Factory(t) { return new (t || ColumnResizeNotifierSource)(); };
ColumnResizeNotifierSource.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ColumnResizeNotifierSource, factory: ColumnResizeNotifierSource.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ColumnResizeNotifierSource, [{
        type: Injectable
    }], function () { return []; }, null); })();
/** Service for triggering column resizes imperatively or being notified of them. */
export class ColumnResizeNotifier {
    constructor(_source) {
        this._source = _source;
        /** Emits whenever a column is resized. */
        this.resizeCompleted = this._source.resizeCompleted;
    }
    /** Instantly resizes the specified column. */
    resize(columnId, size) {
        this._source.triggerResize.next({ columnId, size, completeImmediately: true, isStickyColumn: true });
    }
}
ColumnResizeNotifier.ɵfac = function ColumnResizeNotifier_Factory(t) { return new (t || ColumnResizeNotifier)(ɵngcc0.ɵɵinject(ColumnResizeNotifierSource)); };
ColumnResizeNotifier.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ColumnResizeNotifier, factory: ColumnResizeNotifier.ɵfac });
ColumnResizeNotifier.ctorParameters = () => [
    { type: ColumnResizeNotifierSource }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ColumnResizeNotifier, [{
        type: Injectable
    }], function () { return [{ type: ColumnResizeNotifierSource }]; }, null); })();
