// @angular/cdk-experimental/esm2015/table-scroll-container/table-scroll-container-module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { CdkTableScrollContainer } from './table-scroll-container';
import * as ɵngcc0 from '@angular/core';
export class CdkTableScrollContainerModule {
}
CdkTableScrollContainerModule.ɵfac = function CdkTableScrollContainerModule_Factory(t) { return new (t || CdkTableScrollContainerModule)(); };
CdkTableScrollContainerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CdkTableScrollContainerModule });
CdkTableScrollContainerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkTableScrollContainerModule, [{
        type: NgModule,
        args: [{
                declarations: [CdkTableScrollContainer],
                exports: [CdkTableScrollContainer]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkTableScrollContainerModule, { declarations: function () { return [CdkTableScrollContainer]; }, exports: function () { return [CdkTableScrollContainer]; } }); })();
