// @angular/cdk-experimental/esm2015/column-resize/column-resize-directives/column-resize-flex.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, ElementRef, NgZone } from '@angular/core';
import { CdkTable } from '@angular/cdk/table';
import { ColumnResize } from '../column-resize';
import { ColumnResizeNotifier, ColumnResizeNotifierSource } from '../column-resize-notifier';
import { HeaderRowEventDispatcher } from '../event-dispatcher';
import { FLEX_PROVIDERS } from './constants';
/**
 * Explicitly enables column resizing for a flexbox-based cdk-table.
 * Individual columns must be annotated specifically.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../column-resize-notifier';
import * as ɵngcc2 from '../event-dispatcher';
import * as ɵngcc3 from '@angular/cdk/table';
export class CdkColumnResizeFlex extends ColumnResize {
    constructor(columnResizeNotifier, elementRef, eventDispatcher, ngZone, notifier, table) {
        super();
        this.columnResizeNotifier = columnResizeNotifier;
        this.elementRef = elementRef;
        this.eventDispatcher = eventDispatcher;
        this.ngZone = ngZone;
        this.notifier = notifier;
        this.table = table;
    }
}
CdkColumnResizeFlex.ɵfac = function CdkColumnResizeFlex_Factory(t) { return new (t || CdkColumnResizeFlex)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ColumnResizeNotifier), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.HeaderRowEventDispatcher), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ColumnResizeNotifierSource), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.CdkTable)); };
CdkColumnResizeFlex.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CdkColumnResizeFlex, selectors: [["cdk-table", "columnResize", ""]], features: [ɵngcc0.ɵɵProvidersFeature([
            ...FLEX_PROVIDERS,
            { provide: ColumnResize, useExisting: CdkColumnResizeFlex },
        ]), ɵngcc0.ɵɵInheritDefinitionFeature] });
CdkColumnResizeFlex.ctorParameters = () => [
    { type: ColumnResizeNotifier },
    { type: ElementRef },
    { type: HeaderRowEventDispatcher },
    { type: NgZone },
    { type: ColumnResizeNotifierSource },
    { type: CdkTable }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkColumnResizeFlex, [{
        type: Directive,
        args: [{
                selector: 'cdk-table[columnResize]',
                providers: [
                    ...FLEX_PROVIDERS,
                    { provide: ColumnResize, useExisting: CdkColumnResizeFlex },
                ]
            }]
    }], function () { return [{ type: ɵngcc1.ColumnResizeNotifier }, { type: ɵngcc0.ElementRef }, { type: ɵngcc2.HeaderRowEventDispatcher }, { type: ɵngcc0.NgZone }, { type: ɵngcc1.ColumnResizeNotifierSource }, { type: ɵngcc3.CdkTable }]; }, null); })();
