// kubeflow/esm2015/lib/snack-bar/component/snack-bar.component.js
import { Component, Inject } from '@angular/core';
import { MAT_SNACK_BAR_DATA } from '@angular/material/snack-bar';
import { SnackType } from '../types';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/snack-bar";
import * as i2 from "@angular/material/icon";
import * as i3 from "@angular/common";
import * as i4 from "@angular/material/button";
const _c2 = function (a0) { return [a0, "pad"]; };
export class SnackBarComponent {
    constructor(snackBarRef, data) {
        this.snackBarRef = snackBarRef;
        this.data = data;
    }
    get icon() {
        switch (this.data.snackType) {
            case SnackType.Success:
                return 'done';
            case SnackType.Error:
                return 'clear';
            case SnackType.Warning:
                return 'warning';
            case SnackType.Info:
                return 'info';
            default:
                return 'warning';
        }
    }
    dismiss() {
        this.snackBarRef.dismiss();
    }
}
SnackBarComponent.ɵfac = function SnackBarComponent_Factory(t) { return new (t || SnackBarComponent)(i0.ɵɵdirectiveInject(i1.MatSnackBarRef), i0.ɵɵdirectiveInject(MAT_SNACK_BAR_DATA)); };
SnackBarComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: SnackBarComponent, selectors: [["lib-snack-bar"]], decls: 7, vars: 5, consts: function () { let i18n_0; if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
        const MSG_EXTERNAL_1808188406576936132$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_SNACK_BAR_COMPONENT_SNACK_BAR_COMPONENT_TS_1 = goog.getMsg("DISMISS");
        i18n_0 = MSG_EXTERNAL_1808188406576936132$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_SNACK_BAR_COMPONENT_SNACK_BAR_COMPONENT_TS_1;
    }
    else {
        i18n_0 = $localize `:␟0c0be57aaf30414531e79157a45e0fc3fd6b637b␟1808188406576936132:DISMISS`;
    } return [[1, "snack-container"], [3, "ngClass"], ["mat-button", "", "color", "accent", 3, "click"], i18n_0]; }, template: function SnackBarComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelementStart(1, "mat-icon", 1);
        i0.ɵɵtext(2);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(3, "span");
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(5, "button", 2);
        i0.ɵɵlistener("click", function SnackBarComponent_Template_button_click_5_listener() { return ctx.dismiss(); });
        i0.ɵɵi18n(6, 3);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(3, _c2, ctx.icon));
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx.icon);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate(ctx.data.msg);
    } }, directives: [i2.MatIcon, i3.NgClass, i4.MatButton], styles: [".snack-container[_ngcontent-%COMP%]{display:flex;align-items:center}.pad[_ngcontent-%COMP%]{margin-right:10px}.done[_ngcontent-%COMP%]{color:green}.clear[_ngcontent-%COMP%]{color:red}.warning[_ngcontent-%COMP%]{color:orange}span[_ngcontent-%COMP%]{width:90%}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(SnackBarComponent, [{
        type: Component,
        args: [{
                selector: 'lib-snack-bar',
                templateUrl: './snack-bar.component.html',
                styleUrls: ['./snack-bar.component.scss'],
            }]
    }], function () { return [{ type: i1.MatSnackBarRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [MAT_SNACK_BAR_DATA]
            }] }]; }, null); })();
