// kubeflow/esm2015/lib/title-actions-toolbar/title-actions-toolbar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TitleActionsToolbarComponent } from './title-actions-toolbar.component';
import { MatIconModule } from '@angular/material/icon';
import { MatDividerModule } from '@angular/material/divider';
import { MatButtonModule } from '@angular/material/button';
import * as i0 from "@angular/core";
export class TitleActionsToolbarModule {
}
TitleActionsToolbarModule.ɵfac = function TitleActionsToolbarModule_Factory(t) { return new (t || TitleActionsToolbarModule)(); };
TitleActionsToolbarModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TitleActionsToolbarModule });
TitleActionsToolbarModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, MatIconModule, MatDividerModule, MatButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TitleActionsToolbarModule, [{
        type: NgModule,
        args: [{
                declarations: [TitleActionsToolbarComponent],
                imports: [CommonModule, MatIconModule, MatDividerModule, MatButtonModule],
                exports: [TitleActionsToolbarComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TitleActionsToolbarModule, { declarations: [TitleActionsToolbarComponent], imports: [CommonModule, MatIconModule, MatDividerModule, MatButtonModule], exports: [TitleActionsToolbarComponent] }); })();
