// kubeflow/esm2015/lib/title-actions-toolbar/title-actions-toolbar.component.js
import { Component, Input, EventEmitter, Output, HostBinding, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/divider";
import * as i3 from "@angular/material/button";
import * as i4 from "@angular/material/icon";
function TitleActionsToolbarComponent_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 8);
    i0.ɵɵlistener("click", function TitleActionsToolbarComponent_button_2_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.emitBack(); });
    i0.ɵɵelementStart(1, "mat-icon");
    i0.ɵɵtext(2, "keyboard_backspace");
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} }
function TitleActionsToolbarComponent_ng_container_8_ng_container_1_mat_icon_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "mat-icon", 12);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const button_r4 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", button_r4.icon, " ");
} }
function TitleActionsToolbarComponent_ng_container_8_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r11 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "button", 10);
    i0.ɵɵlistener("click", function TitleActionsToolbarComponent_ng_container_8_ng_container_1_Template_button_click_1_listener() { i0.ɵɵrestoreView(_r11); const button_r4 = i0.ɵɵnextContext().$implicit; return button_r4.fn(); });
    i0.ɵɵtemplate(2, TitleActionsToolbarComponent_ng_container_8_ng_container_1_mat_icon_2_Template, 2, 1, "mat-icon", 11);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const button_r4 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("color", button_r4.color)("disabled", button_r4.disabled);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", button_r4.icon);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", button_r4.text, " ");
} }
function TitleActionsToolbarComponent_ng_container_8_ng_container_2_mat_icon_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "mat-icon", 12);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const button_r4 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", button_r4.icon, " ");
} }
function TitleActionsToolbarComponent_ng_container_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r17 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵelementStart(1, "button", 13);
    i0.ɵɵlistener("click", function TitleActionsToolbarComponent_ng_container_8_ng_container_2_Template_button_click_1_listener() { i0.ɵɵrestoreView(_r17); const button_r4 = i0.ɵɵnextContext().$implicit; return button_r4.fn(); });
    i0.ɵɵtemplate(2, TitleActionsToolbarComponent_ng_container_8_ng_container_2_mat_icon_2_Template, 2, 1, "mat-icon", 11);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const button_r4 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("color", button_r4.color)("disabled", button_r4.disabled);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", button_r4.icon);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", button_r4.text, " ");
} }
function TitleActionsToolbarComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, TitleActionsToolbarComponent_ng_container_8_ng_container_1_Template, 4, 4, "ng-container", 9);
    i0.ɵɵtemplate(2, TitleActionsToolbarComponent_ng_container_8_ng_container_2_Template, 4, 4, "ng-container", 9);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const button_r4 = ctx.$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", button_r4.raised);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", button_r4.stroked);
} }
const _c0 = ["*"];
export class TitleActionsToolbarComponent {
    constructor() {
        this.buttons = [];
        this.backButton = false;
        this.back = new EventEmitter();
        this.selfClass = true;
    }
    emitBack() {
        this.back.emit('backButton');
    }
}
TitleActionsToolbarComponent.ɵfac = function TitleActionsToolbarComponent_Factory(t) { return new (t || TitleActionsToolbarComponent)(); };
TitleActionsToolbarComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: TitleActionsToolbarComponent, selectors: [["lib-title-actions-toolbar"]], hostVars: 2, hostBindings: function TitleActionsToolbarComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("lib-title-actions-toolbar", ctx.selfClass);
    } }, inputs: { buttons: "buttons", backButton: "backButton", title: "title" }, outputs: { back: "back" }, ngContentSelectors: _c0, decls: 10, vars: 3, consts: [[1, "flex"], [1, "page-padding-left"], ["mat-icon-button", "", "color", "primary", "class", "back-button", 3, "click", 4, "ngIf"], [1, "title-margin", "title"], [1, "margin-content"], [1, "toolbar-buttons"], [4, "ngFor", "ngForOf"], [1, "page-placement", "margin-top"], ["mat-icon-button", "", "color", "primary", 1, "back-button", 3, "click"], [4, "ngIf"], ["mat-button", "", 1, "toolbar-button", 3, "color", "disabled", "click"], ["class", "button-icon", 4, "ngIf"], [1, "button-icon"], ["mat-stroked-button", "", 1, "toolbar-button", 3, "color", "disabled", "click"]], template: function TitleActionsToolbarComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelement(1, "div", 1);
        i0.ɵɵtemplate(2, TitleActionsToolbarComponent_button_2_Template, 3, 0, "button", 2);
        i0.ɵɵelementStart(3, "div", 3);
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(5, "div", 4);
        i0.ɵɵprojection(6);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(7, "div", 5);
        i0.ɵɵtemplate(8, TitleActionsToolbarComponent_ng_container_8_Template, 3, 2, "ng-container", 6);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelement(9, "mat-divider", 7);
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.backButton);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx.title, " ");
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngForOf", ctx.buttons);
    } }, directives: [i1.NgIf, i1.NgForOf, i2.MatDivider, i3.MatButton, i4.MatIcon], styles: ["[_nghost-%COMP%]{display:block;width:100%;padding-top:8px}.title[_ngcontent-%COMP%]{font-weight:400;font-size:20px}.back-button[_ngcontent-%COMP%]{margin:auto 0}.actions-wrapper[_ngcontent-%COMP%]{margin-top:.22rem}.title-margin[_ngcontent-%COMP%]{margin:auto 0}.button-icon[_ngcontent-%COMP%]{font-size:1.2rem;padding-top:.1rem}.padding-bottom[_ngcontent-%COMP%]{padding-bottom:2rem}.margin-top[_ngcontent-%COMP%]{margin-top:.2rem}.margin-content[_ngcontent-%COMP%]{margin:auto}.toolbar-buttons[_ngcontent-%COMP%]{margin:auto 0}.toolbar-button[_ngcontent-%COMP%]{margin-left:.2rem}.toolbar-buttons[_ngcontent-%COMP%]   .toolbar-button[_ngcontent-%COMP%]:last-child{margin-right:20px}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TitleActionsToolbarComponent, [{
        type: Component,
        args: [{
                selector: 'lib-title-actions-toolbar',
                templateUrl: './title-actions-toolbar.component.html',
                styleUrls: ['./title-actions-toolbar.component.scss'],
            }]
    }], null, { buttons: [{
            type: Input
        }], backButton: [{
            type: Input
        }], title: [{
            type: Input
        }], back: [{
            type: Output
        }], selfClass: [{
            type: HostBinding,
            args: ['class.lib-title-actions-toolbar']
        }] }); })();
