// @angular/cdk-experimental/esm2015/menu/menu-group.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, Output, EventEmitter, ContentChildren, QueryList, } from '@angular/core';
import { UniqueSelectionDispatcher } from '@angular/cdk/collections';
import { takeUntil } from 'rxjs/operators';
import { CdkMenuItemSelectable } from './menu-item-selectable';
/**
 * Directive which acts as a grouping container for `CdkMenuItem` instances with
 * `role="menuitemradio"`, similar to a `role="radiogroup"` element.
 */
import * as ɵngcc0 from '@angular/core';
export class CdkMenuGroup {
    constructor() {
        /** Emits the element when checkbox or radiobutton state changed  */
        this.change = new EventEmitter();
        /** Emits when the _selectableItems QueryList triggers a change */
        this._selectableChanges = new EventEmitter();
    }
    ngAfterContentInit() {
        this._registerMenuSelectionListeners();
    }
    /**
     * Register the child selectable elements with the change emitter and ensure any new child
     * elements do so as well.
     */
    _registerMenuSelectionListeners() {
        this._selectableItems.forEach(selectable => this._registerClickListener(selectable));
        this._selectableItems.changes.subscribe((selectableItems) => {
            this._selectableChanges.next();
            selectableItems.forEach(selectable => this._registerClickListener(selectable));
        });
    }
    /** Register each selectable to emit on the change Emitter when clicked */
    _registerClickListener(selectable) {
        selectable.toggled
            .pipe(takeUntil(this._selectableChanges))
            .subscribe(() => this.change.next(selectable));
    }
    ngOnDestroy() {
        this._selectableChanges.next();
        this._selectableChanges.complete();
    }
}
CdkMenuGroup.ɵfac = function CdkMenuGroup_Factory(t) { return new (t || CdkMenuGroup)(); };
CdkMenuGroup.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CdkMenuGroup, selectors: [["", "cdkMenuGroup", ""]], contentQueries: function CdkMenuGroup_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, CdkMenuItemSelectable, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._selectableItems = _t);
    } }, hostAttrs: ["role", "group", 1, "cdk-menu-group"], outputs: { change: "change" }, exportAs: ["cdkMenuGroup"], features: [ɵngcc0.ɵɵProvidersFeature([{ provide: UniqueSelectionDispatcher, useClass: UniqueSelectionDispatcher }])] });
CdkMenuGroup.propDecorators = {
    change: [{ type: Output }],
    _selectableItems: [{ type: ContentChildren, args: [CdkMenuItemSelectable, { descendants: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkMenuGroup, [{
        type: Directive,
        args: [{
                selector: '[cdkMenuGroup]',
                exportAs: 'cdkMenuGroup',
                host: {
                    'role': 'group',
                    'class': 'cdk-menu-group'
                },
                providers: [{ provide: UniqueSelectionDispatcher, useClass: UniqueSelectionDispatcher }]
            }]
    }], function () { return []; }, { change: [{
            type: Output
        }], _selectableItems: [{
            type: ContentChildren,
            args: [CdkMenuItemSelectable, { descendants: true }]
        }] }); })();
