// @angular/cdk-experimental/esm2015/menu/menu-item-selectable.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Input, Directive, Output, EventEmitter } from '@angular/core';
import { CdkMenuItem } from './menu-item';
/** Counter used to set a unique id and name for a selectable item */
import * as ɵngcc0 from '@angular/core';
let nextId = 0;
/**
 * Base class providing checked state for MenuItems along with outputting a clicked event when the
 * element is triggered. It provides functionality for selectable elements.
 */
export class CdkMenuItemSelectable extends CdkMenuItem {
    constructor() {
        super(...arguments);
        /** Event emitted when the selectable item is clicked */
        this.toggled = new EventEmitter();
        this._checked = false;
        /** The name of the selectable element with a default value */
        this.name = `cdk-selectable-item-${nextId++}`;
        /** The id of the selectable element with a default value */
        this.id = `cdk-selectable-item-${nextId++}`;
    }
    /** Whether the element is checked */
    get checked() {
        return this._checked;
    }
    set checked(value) {
        this._checked = coerceBooleanProperty(value);
    }
    /** If the element is not disabled emit the click event */
    trigger() {
        if (!this.disabled) {
            this.toggled.next(this);
        }
    }
}
CdkMenuItemSelectable.ɵfac = /*@__PURE__*/ function () { let ɵCdkMenuItemSelectable_BaseFactory; return function CdkMenuItemSelectable_Factory(t) { return (ɵCdkMenuItemSelectable_BaseFactory || (ɵCdkMenuItemSelectable_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(CdkMenuItemSelectable)))(t || CdkMenuItemSelectable); }; }();
CdkMenuItemSelectable.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CdkMenuItemSelectable, inputs: { name: "name", id: "id", checked: "checked" }, outputs: { toggled: "cdkMenuItemToggled" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
CdkMenuItemSelectable.propDecorators = {
    toggled: [{ type: Output, args: ['cdkMenuItemToggled',] }],
    checked: [{ type: Input }],
    name: [{ type: Input }],
    id: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkMenuItemSelectable, [{
        type: Directive
    }], null, { toggled: [{
            type: Output,
            args: ['cdkMenuItemToggled']
        }], name: [{
            type: Input
        }], id: [{
            type: Input
        }], checked: [{
            type: Input
        }] }); })();
