// @angular/cdk-experimental/esm2015/selection/row-selection.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { Directive, Input } from '@angular/core';
import { CdkSelection } from './selection';
/**
 * Applies `cdk-selected` class and `aria-selected` to an element.
 *
 * Must be used within a parent `CdkSelection` directive.
 * Must be provided with the value. The index is required if `trackBy` is used on the `CdkSelection`
 * directive.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './selection';
export class CdkRowSelection {
    constructor(_selection) {
        this._selection = _selection;
    }
    get index() { return this._index; }
    set index(index) { this._index = coerceNumberProperty(index); }
}
CdkRowSelection.ɵfac = function CdkRowSelection_Factory(t) { return new (t || CdkRowSelection)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CdkSelection)); };
CdkRowSelection.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CdkRowSelection, selectors: [["", "cdkRowSelection", ""]], hostVars: 3, hostBindings: function CdkRowSelection_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-selected", ctx._selection.isSelected(ctx.value, ctx.index));
        ɵngcc0.ɵɵclassProp("cdk-selected", ctx._selection.isSelected(ctx.value, ctx.index));
    } }, inputs: { index: ["cdkRowSelectionIndex", "index"], value: ["cdkRowSelectionValue", "value"] } });
CdkRowSelection.ctorParameters = () => [
    { type: CdkSelection }
];
CdkRowSelection.propDecorators = {
    value: [{ type: Input, args: ['cdkRowSelectionValue',] }],
    index: [{ type: Input, args: ['cdkRowSelectionIndex',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkRowSelection, [{
        type: Directive,
        args: [{
                selector: '[cdkRowSelection]',
                host: {
                    '[class.cdk-selected]': '_selection.isSelected(this.value, this.index)',
                    '[attr.aria-selected]': '_selection.isSelected(this.value, this.index)'
                }
            }]
    }], function () { return [{ type: ɵngcc1.CdkSelection }]; }, { index: [{
            type: Input,
            args: ['cdkRowSelectionIndex']
        }], value: [{
            type: Input,
            args: ['cdkRowSelectionValue']
        }] }); })();
