// @angular/cdk-experimental/esm2015/selection/selection-column.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { CdkCellDef, CdkColumnDef, CdkHeaderCellDef, CdkTable } from '@angular/cdk/table';
import { Component, Input, Optional, ViewChild, ChangeDetectionStrategy, ViewEncapsulation, Inject, } from '@angular/core';
import { CdkSelection } from './selection';
/**
 * Column that adds row selecting checkboxes and a select-all checkbox if `cdkSelectionMultiple` is
 * `true`.
 *
 * Must be used within a parent `CdkSelection` directive.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/table';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './select-all';
import * as ɵngcc4 from './selection-toggle';
import * as ɵngcc5 from './selection';

function CdkSelectionColumn_th_1_input_1_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 5, 6);
    ɵngcc0.ɵɵlistener("click", function CdkSelectionColumn_th_1_input_1_Template_input_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const _r3 = ɵngcc0.ɵɵreference(1); return _r3.toggle($event); });
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r3 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵproperty("checked", ɵngcc0.ɵɵpipeBind1(2, 2, _r3.checked))("indeterminate", ɵngcc0.ɵɵpipeBind1(3, 4, _r3.indeterminate));
} }
function CdkSelectionColumn_th_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 3);
    ɵngcc0.ɵɵtemplate(1, CdkSelectionColumn_th_1_input_1_Template, 4, 6, "input", 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.selection.multiple);
} }
function CdkSelectionColumn_td_2_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "td", 7);
    ɵngcc0.ɵɵelementStart(1, "input", 8, 9);
    ɵngcc0.ɵɵlistener("click", function CdkSelectionColumn_td_2_Template_input_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r10); const _r8 = ɵngcc0.ɵɵreference(2); return _r8.toggle(); });
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const row_r6 = ctx.$implicit;
    const i_r7 = ctx.$index;
    const _r8 = ɵngcc0.ɵɵreference(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkSelectionToggleValue", row_r6)("cdkSelectionToggleIndex", i_r7)("checked", ɵngcc0.ɵɵpipeBind1(3, 3, _r8.checked));
} }
export class CdkSelectionColumn {
    constructor(_table, selection) {
        this._table = _table;
        this.selection = selection;
    }
    /** Column name that should be used to reference this column. */
    get name() {
        return this._name;
    }
    set name(name) {
        this._name = name;
        this._syncColumnDefName();
    }
    ngOnInit() {
        if (!this.selection && (typeof ngDevMode === 'undefined' || ngDevMode)) {
            throw Error('CdkSelectionColumn: missing CdkSelection in the parent');
        }
        this._syncColumnDefName();
        if (this._table) {
            this._columnDef.cell = this._cell;
            this._columnDef.headerCell = this._headerCell;
            this._table.addColumnDef(this._columnDef);
        }
        else if ((typeof ngDevMode === 'undefined' || ngDevMode)) {
            throw Error('CdkSelectionColumn: missing parent table');
        }
    }
    ngOnDestroy() {
        if (this._table) {
            this._table.removeColumnDef(this._columnDef);
        }
    }
    _syncColumnDefName() {
        if (this._columnDef) {
            this._columnDef.name = this._name;
        }
    }
}
CdkSelectionColumn.ɵfac = function CdkSelectionColumn_Factory(t) { return new (t || CdkSelectionColumn)(ɵngcc0.ɵɵdirectiveInject(CdkTable, 8), ɵngcc0.ɵɵdirectiveInject(CdkSelection, 8)); };
CdkSelectionColumn.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CdkSelectionColumn, selectors: [["cdk-selection-column"]], viewQuery: function CdkSelectionColumn_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(CdkColumnDef, 7);
        ɵngcc0.ɵɵviewQuery(CdkCellDef, 7);
        ɵngcc0.ɵɵviewQuery(CdkHeaderCellDef, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._columnDef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._cell = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._headerCell = _t.first);
    } }, inputs: { name: ["cdkSelectionColumnName", "name"] }, decls: 3, vars: 0, consts: [["cdkColumnDef", ""], ["cdkHeaderCell", "", 4, "cdkHeaderCellDef"], ["cdkCell", "", 4, "cdkCellDef"], ["cdkHeaderCell", ""], ["type", "checkbox", "cdkSelectAll", "", 3, "checked", "indeterminate", "click", 4, "ngIf"], ["type", "checkbox", "cdkSelectAll", "", 3, "checked", "indeterminate", "click"], ["allToggler", "cdkSelectAll"], ["cdkCell", ""], ["type", "checkbox", "cdkSelectionToggle", "", 3, "cdkSelectionToggleValue", "cdkSelectionToggleIndex", "checked", "click"], ["toggler", "cdkSelectionToggle"]], template: function CdkSelectionColumn_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0, 0);
        ɵngcc0.ɵɵtemplate(1, CdkSelectionColumn_th_1_Template, 2, 1, "th", 1);
        ɵngcc0.ɵɵtemplate(2, CdkSelectionColumn_td_2_Template, 4, 5, "td", 2);
        ɵngcc0.ɵɵelementContainerEnd();
    } }, directives: [ɵngcc1.CdkColumnDef, ɵngcc1.CdkHeaderCellDef, ɵngcc1.CdkCellDef, ɵngcc2.NgIf, ɵngcc3.CdkSelectAll, ɵngcc4.CdkSelectionToggle], pipes: [ɵngcc2.AsyncPipe], encapsulation: 2, changeDetection: 0 });
CdkSelectionColumn.ctorParameters = () => [
    { type: CdkTable, decorators: [{ type: Optional }, { type: Inject, args: [CdkTable,] }] },
    { type: CdkSelection, decorators: [{ type: Optional }, { type: Inject, args: [CdkSelection,] }] }
];
CdkSelectionColumn.propDecorators = {
    name: [{ type: Input, args: ['cdkSelectionColumnName',] }],
    _columnDef: [{ type: ViewChild, args: [CdkColumnDef, { static: true },] }],
    _cell: [{ type: ViewChild, args: [CdkCellDef, { static: true },] }],
    _headerCell: [{ type: ViewChild, args: [CdkHeaderCellDef, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkSelectionColumn, [{
        type: Component,
        args: [{
                selector: 'cdk-selection-column',
                template: `
    <ng-container cdkColumnDef>
      <th cdkHeaderCell *cdkHeaderCellDef>
        <input type="checkbox" *ngIf="selection.multiple"
            cdkSelectAll
            #allToggler="cdkSelectAll"
            [checked]="allToggler.checked | async"
            [indeterminate]="allToggler.indeterminate | async"
            (click)="allToggler.toggle($event)">
      </th>
      <td cdkCell *cdkCellDef="let row; let i = $index">
        <input type="checkbox"
            #toggler="cdkSelectionToggle"
            cdkSelectionToggle
            [cdkSelectionToggleValue]="row"
            [cdkSelectionToggleIndex]="i"
            (click)="toggler.toggle()"
            [checked]="toggler.checked | async">
      </td>
    </ng-container>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: ɵngcc1.CdkTable, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [CdkTable]
            }] }, { type: ɵngcc5.CdkSelection, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [CdkSelection]
            }] }]; }, { name: [{
            type: Input,
            args: ['cdkSelectionColumnName']
        }], _columnDef: [{
            type: ViewChild,
            args: [CdkColumnDef, { static: true }]
        }], _cell: [{
            type: ViewChild,
            args: [CdkCellDef, { static: true }]
        }], _headerCell: [{
            type: ViewChild,
            args: [CdkHeaderCellDef, { static: true }]
        }] }); })();
