// @angular/cdk-experimental/esm2015/selection/selection-module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { CdkTableModule } from '@angular/cdk/table';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CdkRowSelection } from './row-selection';
import { CdkSelectAll } from './select-all';
import { CdkSelection } from './selection';
import { CdkSelectionColumn } from './selection-column';
import { CdkSelectionToggle } from './selection-toggle';
import * as ɵngcc0 from '@angular/core';
export class CdkSelectionModule {
}
CdkSelectionModule.ɵfac = function CdkSelectionModule_Factory(t) { return new (t || CdkSelectionModule)(); };
CdkSelectionModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CdkSelectionModule });
CdkSelectionModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            CdkTableModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkSelectionModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    CdkTableModule,
                ],
                exports: [
                    CdkSelection,
                    CdkSelectionToggle,
                    CdkSelectAll,
                    CdkSelectionColumn,
                    CdkRowSelection,
                ],
                declarations: [
                    CdkSelection,
                    CdkSelectionToggle,
                    CdkSelectAll,
                    CdkSelectionColumn,
                    CdkRowSelection,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkSelectionModule, { declarations: function () { return [CdkSelection,
        CdkSelectionToggle,
        CdkSelectAll,
        CdkSelectionColumn,
        CdkRowSelection]; }, imports: function () { return [CommonModule,
        CdkTableModule]; }, exports: function () { return [CdkSelection,
        CdkSelectionToggle,
        CdkSelectAll,
        CdkSelectionColumn,
        CdkRowSelection]; } }); })();
