// kubeflow/esm2015/lib/panel/panel.component.js
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/icon";
const _c0 = ["*"];
export class PanelComponent {
    constructor() {
        this.icon = 'info';
        this.color = 'primary';
    }
}
PanelComponent.ɵfac = function PanelComponent_Factory(t) { return new (t || PanelComponent)(); };
PanelComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: PanelComponent, selectors: [["lib-panel"]], inputs: { icon: "icon", color: "color", message: "message" }, ngContentSelectors: _c0, decls: 6, vars: 3, consts: [[1, "panel-body", "flex"], [1, "panel-icon", 3, "color"], [1, "panel-message"]], template: function PanelComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelementStart(1, "mat-icon", 1);
        i0.ɵɵtext(2);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(3, "div", 2);
        i0.ɵɵtext(4);
        i0.ɵɵprojection(5);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("color", ctx.color);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx.icon);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate1(" ", ctx.message, " ");
    } }, directives: [i1.MatIcon], styles: [".panel-body[_ngcontent-%COMP%]{background-color:rgba(0,0,0,.05);border-radius:4px;padding:1px;display:flex}.panel-icon[_ngcontent-%COMP%]{display:inline-block;margin:16px 24px}.panel-message[_ngcontent-%COMP%]{flex-wrap:wrap;margin-top:auto;margin-bottom:auto;margin-right:1rem}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(PanelComponent, [{
        type: Component,
        args: [{
                selector: 'lib-panel',
                templateUrl: './panel.component.html',
                styleUrls: ['./panel.component.scss'],
            }]
    }], null, { icon: [{
            type: Input
        }], color: [{
            type: Input
        }], message: [{
            type: Input
        }] }); })();
