// @angular/cdk-experimental/esm2015/combobox/combobox-module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgModule } from '@angular/core';
import { OverlayModule } from '@angular/cdk/overlay';
import { CdkCombobox } from './combobox';
import { CdkComboboxPanel } from './combobox-panel';
import { CdkComboboxPopup } from './combobox-popup';
import * as ɵngcc0 from '@angular/core';
const EXPORTED_DECLARATIONS = [CdkCombobox, CdkComboboxPanel, CdkComboboxPopup];
export class CdkComboboxModule {
}
CdkComboboxModule.ɵfac = function CdkComboboxModule_Factory(t) { return new (t || CdkComboboxModule)(); };
CdkComboboxModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: CdkComboboxModule });
CdkComboboxModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[OverlayModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkComboboxModule, [{
        type: NgModule,
        args: [{
                imports: [OverlayModule],
                exports: EXPORTED_DECLARATIONS,
                declarations: EXPORTED_DECLARATIONS
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CdkComboboxModule, { declarations: function () { return [CdkCombobox, CdkComboboxPanel, CdkComboboxPopup]; }, imports: function () { return [OverlayModule]; }, exports: function () { return [CdkCombobox, CdkComboboxPanel, CdkComboboxPopup]; } }); })();
