// @angular/cdk-experimental/esm2015/combobox/combobox-popup.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Directive, ElementRef, Inject, InjectionToken, Input, Optional } from '@angular/core';
import { CdkComboboxPanel } from './combobox-panel';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './combobox-panel';
export const PANEL = new InjectionToken('CdkComboboxPanel');
let nextId = 0;
export class CdkComboboxPopup {
    constructor(_elementRef, _parentPanel) {
        this._elementRef = _elementRef;
        this._parentPanel = _parentPanel;
        this._role = 'dialog';
        this.id = `cdk-combobox-popup-${nextId++}`;
    }
    get role() {
        return this._role;
    }
    set role(value) {
        this._role = value;
    }
    get firstFocus() {
        return this._firstFocusElement;
    }
    set firstFocus(id) {
        this._firstFocusElement = id;
    }
    ngOnInit() {
        this.registerWithPanel();
    }
    registerWithPanel() {
        if (this._parentPanel === null || this._parentPanel === undefined) {
            this._explicitPanel._registerContent(this.id, this._role);
        }
        else {
            this._parentPanel._registerContent(this.id, this._role);
        }
    }
    focusFirstElement() {
        if (this._firstFocusElement) {
            this._firstFocusElement.focus();
        }
        else {
            this._elementRef.nativeElement.focus();
        }
    }
}
CdkComboboxPopup.ɵfac = function CdkComboboxPopup_Factory(t) { return new (t || CdkComboboxPopup)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(PANEL, 8)); };
CdkComboboxPopup.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CdkComboboxPopup, selectors: [["", "cdkComboboxPopup", ""]], hostAttrs: ["tabindex", "-1", 1, "cdk-combobox-popup"], hostVars: 2, hostBindings: function CdkComboboxPopup_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("focus", function CdkComboboxPopup_focus_HostBindingHandler() { return ctx.focusFirstElement(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("id", ctx.id);
        ɵngcc0.ɵɵattribute("role", ctx.role);
    } }, inputs: { id: "id", role: "role", firstFocus: "firstFocus", _explicitPanel: ["parentPanel", "_explicitPanel"] }, exportAs: ["cdkComboboxPopup"] });
CdkComboboxPopup.ctorParameters = () => [
    { type: ElementRef },
    { type: CdkComboboxPanel, decorators: [{ type: Optional }, { type: Inject, args: [PANEL,] }] }
];
CdkComboboxPopup.propDecorators = {
    role: [{ type: Input }],
    firstFocus: [{ type: Input }],
    id: [{ type: Input }],
    _explicitPanel: [{ type: Input, args: ['parentPanel',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CdkComboboxPopup, [{
        type: Directive,
        args: [{
                selector: '[cdkComboboxPopup]',
                exportAs: 'cdkComboboxPopup',
                host: {
                    'class': 'cdk-combobox-popup',
                    '[attr.role]': 'role',
                    '[id]': 'id',
                    'tabindex': '-1',
                    '(focus)': 'focusFirstElement()'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.CdkComboboxPanel, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [PANEL]
            }] }]; }, { id: [{
            type: Input
        }], role: [{
            type: Input
        }], firstFocus: [{
            type: Input
        }], _explicitPanel: [{
            type: Input,
            args: ['parentPanel']
        }] }); })();
