// kubeflow/esm2015/lib/kubeflow.module.js
import '@angular/localize/init';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NamespaceSelectModule } from './namespace-select/namespace-select.module';
import { ResourceTableModule } from './resource-table/resource-table.module';
import { SnackBarModule } from './snack-bar/snack-bar.module';
import { FormModule } from './form/form.module';
import { HttpClientModule, HTTP_INTERCEPTORS, HttpClientXsrfModule, } from '@angular/common/http';
import { HeadersInterceptor } from './services/rok/injector';
import { PopoverModule } from './popover/popover.module';
import { TitleActionsToolbarModule } from './title-actions-toolbar/title-actions-toolbar.module';
import { ConditionsTableModule } from './conditions-table/conditions-table.module';
import { DetailsListModule } from './details-list/details-list.module';
import { DateTimeModule } from './date-time/date-time.module';
import { PanelModule } from './panel/panel.module';
import { LoadingSpinnerModule } from './loading-spinner/loading-spinner.module';
import { ConfirmDialogModule } from './confirm-dialog/confirm-dialog.module';
import * as i0 from "@angular/core";
export class KubeflowModule {
}
KubeflowModule.ɵfac = function KubeflowModule_Factory(t) { return new (t || KubeflowModule)(); };
KubeflowModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: KubeflowModule });
KubeflowModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
        { provide: HTTP_INTERCEPTORS, useClass: HeadersInterceptor, multi: true },
    ], imports: [[CommonModule, HttpClientModule, HttpClientXsrfModule], NamespaceSelectModule,
        ResourceTableModule,
        SnackBarModule,
        FormModule,
        PopoverModule,
        ConfirmDialogModule,
        HttpClientModule,
        HttpClientXsrfModule,
        TitleActionsToolbarModule,
        ConditionsTableModule,
        DetailsListModule,
        DateTimeModule,
        PanelModule,
        LoadingSpinnerModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(KubeflowModule, [{
        type: NgModule,
        args: [{
                declarations: [],
                exports: [
                    NamespaceSelectModule,
                    ResourceTableModule,
                    SnackBarModule,
                    FormModule,
                    PopoverModule,
                    ConfirmDialogModule,
                    HttpClientModule,
                    HttpClientXsrfModule,
                    TitleActionsToolbarModule,
                    ConditionsTableModule,
                    DetailsListModule,
                    DateTimeModule,
                    PanelModule,
                    LoadingSpinnerModule,
                ],
                imports: [CommonModule, HttpClientModule, HttpClientXsrfModule],
                providers: [
                    { provide: HTTP_INTERCEPTORS, useClass: HeadersInterceptor, multi: true },
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(KubeflowModule, { imports: [CommonModule, HttpClientModule, HttpClientXsrfModule], exports: [NamespaceSelectModule,
        ResourceTableModule,
        SnackBarModule,
        FormModule,
        PopoverModule,
        ConfirmDialogModule,
        HttpClientModule,
        HttpClientXsrfModule,
        TitleActionsToolbarModule,
        ConditionsTableModule,
        DetailsListModule,
        DateTimeModule,
        PanelModule,
        LoadingSpinnerModule] }); })();
