// kubeflow/esm2015/lib/resource-table/action/action.component.js
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { STATUS_TYPE } from '../status/types';
import { Subject } from 'rxjs';
import { ActionEvent } from '../types';
import { get as getAttributeValue } from 'lodash';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/button";
import * as i3 from "@angular/material/tooltip";
import * as i4 from "../../icon/icon.component";
import * as i5 from "@angular/material/progress-spinner";
function ActionComponent_button_0_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 3);
    i0.ɵɵlistener("click", function ActionComponent_button_0_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.emitClickedEvent(); });
    i0.ɵɵelement(1, "lib-icon", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r0.action.tooltip);
    i0.ɵɵproperty("color", ctx_r0.action.color);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("icon", ctx_r0.action.iconReady);
} }
function ActionComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 3);
    i0.ɵɵlistener("click", function ActionComponent_button_1_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6.emitClickedEvent(); });
    i0.ɵɵelement(1, "lib-icon", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r1.action.tooltip);
    i0.ɵɵproperty("color", ctx_r1.action.color);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("icon", ctx_r1.action.iconInit);
} }
function ActionComponent_mat_spinner_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "mat-spinner", 5);
} }
function ActionComponent_button_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "button", 6);
    i0.ɵɵelement(1, "lib-icon", 4);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("icon", ctx_r3.action.iconInit);
} }
export class ActionComponent {
    constructor() {
        // READY: Button will be enabled
        // WAITING: Button will be a Spinner
        // TERMINATING/UNAVAILABLE: Button will be disabled
        this.innerData = {};
        this.clicked = false;
        this.cancelWaitingPhase$ = new Subject();
        this.emitter = new EventEmitter();
    }
    ngOnInit() { }
    // Event emitting functions
    emitClickedEvent() {
        const ev = new ActionEvent(this.action.name, this.data);
        this.emitter.emit(ev);
    }
    // Helpers for handling the Tooltips
    get tooltipInit() {
        if (this.action.tooltip && this.action.tooltip.length > 0) {
            return this.action.tooltip;
        }
        return this.action.tooltipInit;
    }
    get tooltipReady() {
        if (this.action.tooltip && this.action.tooltip.length > 0) {
            return this.action.tooltip;
        }
        return this.action.tooltipReady;
    }
    // Icon handling functions
    getIcon(icon) {
        if (icon.split(':').length !== 2) {
            return '';
        }
        if (this.getCategory(icon) === 'fa') {
            const inpt = icon.split(':');
            return inpt.slice(1, inpt.length);
        }
        return icon.split(':')[1];
    }
    getCategory(icon) {
        if (icon.split(':').length !== 2) {
            return '';
        }
        return icon.split(':')[0];
    }
    // Helpers for checking the Action's State
    isPhaseReady() {
        return this.status === STATUS_TYPE.READY;
    }
    isPhaseInit() {
        return this.status === STATUS_TYPE.UNINITIALIZED;
    }
    isPhaseWaiting() {
        return this.status === STATUS_TYPE.WAITING;
    }
    isPhaseDisabled() {
        return (this.status === STATUS_TYPE.TERMINATING ||
            this.status === STATUS_TYPE.UNAVAILABLE);
    }
    get status() {
        const phaseField = this.action.field;
        if (!phaseField) {
            return STATUS_TYPE.READY;
        }
        const status = getAttributeValue(this.data, phaseField);
        return status;
    }
}
ActionComponent.ɵfac = function ActionComponent_Factory(t) { return new (t || ActionComponent)(); };
ActionComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: ActionComponent, selectors: [["lib-action"]], inputs: { action: "action", data: "data" }, outputs: { emitter: "emitter" }, decls: 4, vars: 4, consts: [["mat-icon-button", "", "matTooltipClass", "custom-tooltip", 3, "matTooltip", "color", "click", 4, "ngIf"], ["diameter", "20", 4, "ngIf"], ["mat-icon-button", "", "disabled", "", 4, "ngIf"], ["mat-icon-button", "", "matTooltipClass", "custom-tooltip", 3, "matTooltip", "color", "click"], [3, "icon"], ["diameter", "20"], ["mat-icon-button", "", "disabled", ""]], template: function ActionComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, ActionComponent_button_0_Template, 2, 3, "button", 0);
        i0.ɵɵtemplate(1, ActionComponent_button_1_Template, 2, 3, "button", 0);
        i0.ɵɵtemplate(2, ActionComponent_mat_spinner_2_Template, 1, 0, "mat-spinner", 1);
        i0.ɵɵtemplate(3, ActionComponent_button_3_Template, 2, 1, "button", 2);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.isPhaseReady());
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isPhaseInit());
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isPhaseWaiting());
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isPhaseDisabled());
    } }, directives: [i1.NgIf, i2.MatButton, i3.MatTooltip, i4.IconComponent, i5.MatSpinner], styles: ["mat-spinner[_ngcontent-%COMP%]{margin:auto}.folder-search-button[_ngcontent-%COMP%]{position:relative}.material-icons.search[_ngcontent-%COMP%]{position:absolute;color:#fff;left:13px;top:11px;font-size:14px}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ActionComponent, [{
        type: Component,
        args: [{
                selector: 'lib-action',
                templateUrl: './action.component.html',
                styleUrls: ['./action.component.scss'],
            }]
    }], function () { return []; }, { action: [{
            type: Input
        }], data: [{
            type: Input
        }], emitter: [{
            type: Output
        }] }); })();
