// kubeflow/esm2015/lib/form/name-namespace-inputs/name-namespace-inputs.component.js
import { Component, Input } from '@angular/core';
import { MAX_NAME_LENGTH } from '../validators';
import * as i0 from "@angular/core";
import * as i1 from "./name-input/name-input.component";
import * as i2 from "@angular/material/form-field";
import * as i3 from "@angular/material/input";
import * as i4 from "@angular/forms";
export class NameNamespaceInputsComponent {
    constructor() {
        this.maxLength = MAX_NAME_LENGTH;
    }
    ngOnInit() { }
}
NameNamespaceInputsComponent.ɵfac = function NameNamespaceInputsComponent_Factory(t) { return new (t || NameNamespaceInputsComponent)(); };
NameNamespaceInputsComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: NameNamespaceInputsComponent, selectors: [["lib-form-name-namespace-inputs"]], inputs: { nameControl: "nameControl", namespaceControl: "namespaceControl", resourceName: "resourceName", maxLength: "maxLength", existingNames: "existingNames" }, decls: 6, vars: 5, consts: function () { let i18n_0; if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
        const MSG_EXTERNAL_3294686077659093992$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_FORM_NAME_NAMESPACE_INPUTS_NAME_NAMESPACE_INPUTS_COMPONENT_TS_1 = goog.getMsg("Namespace");
        i18n_0 = MSG_EXTERNAL_3294686077659093992$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_FORM_NAME_NAMESPACE_INPUTS_NAME_NAMESPACE_INPUTS_COMPONENT_TS_1;
    }
    else {
        i18n_0 = $localize `:␟130fd872c78271a8f86b1ab16a76e823969c47d9␟3294686077659093992:Namespace`;
    } let i18n_2; if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
        const MSG_EXTERNAL_3294686077659093992$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_FORM_NAME_NAMESPACE_INPUTS_NAME_NAMESPACE_INPUTS_COMPONENT_TS_3 = goog.getMsg("Namespace");
        i18n_2 = MSG_EXTERNAL_3294686077659093992$$________________________USERS_JACOBNEYER_CHARIOT_NOTEBOOKS_FRONTEND_KUBEFLOW_COMMON_LIB_PROJECTS_KUBEFLOW_SRC_LIB_FORM_NAME_NAMESPACE_INPUTS_NAME_NAMESPACE_INPUTS_COMPONENT_TS_3;
    }
    else {
        i18n_2 = $localize `:␟130fd872c78271a8f86b1ab16a76e823969c47d9␟3294686077659093992:Namespace`;
    } return [[1, "row"], [1, "column", 3, "nameControl", "maxLength", "resourceName", "existingNames"], ["appearance", "outline", 1, "column"], i18n_0, ["matInput", "", "placeholder", i18n_2, "readonly", "", 3, "formControl"]]; }, template: function NameNamespaceInputsComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵelement(1, "lib-name-input", 1);
        i0.ɵɵelementStart(2, "mat-form-field", 2);
        i0.ɵɵelementStart(3, "mat-label");
        i0.ɵɵi18n(4, 3);
        i0.ɵɵelementEnd();
        i0.ɵɵelement(5, "input", 4);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("nameControl", ctx.nameControl)("maxLength", ctx.maxLength)("resourceName", ctx.resourceName)("existingNames", ctx.existingNames);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("formControl", ctx.namespaceControl);
    } }, directives: [i1.NameInputComponent, i2.MatFormField, i2.MatLabel, i3.MatInput, i4.DefaultValueAccessor, i4.NgControlStatus, i4.FormControlDirective], styles: [""] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(NameNamespaceInputsComponent, [{
        type: Component,
        args: [{
                selector: 'lib-form-name-namespace-inputs',
                templateUrl: './name-namespace-inputs.component.html',
                styleUrls: ['./name-namespace-inputs.component.scss'],
            }]
    }], function () { return []; }, { nameControl: [{
            type: Input
        }], namespaceControl: [{
            type: Input
        }], resourceName: [{
            type: Input
        }], maxLength: [{
            type: Input
        }], existingNames: [{
            type: Input
        }] }); })();
