// kubeflow/esm2015/lib/namespace-select/namespace-select.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatSelectModule } from '@angular/material/select';
import { NamespaceSelectComponent } from './namespace-select.component';
import { SnackBarModule } from '../snack-bar/snack-bar.module';
import * as i0 from "@angular/core";
export class NamespaceSelectModule {
}
NamespaceSelectModule.ɵfac = function NamespaceSelectModule_Factory(t) { return new (t || NamespaceSelectModule)(); };
NamespaceSelectModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: NamespaceSelectModule });
NamespaceSelectModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            HttpClientModule,
            BrowserAnimationsModule,
            MatFormFieldModule,
            MatSelectModule,
            SnackBarModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(NamespaceSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    HttpClientModule,
                    BrowserAnimationsModule,
                    MatFormFieldModule,
                    MatSelectModule,
                    SnackBarModule,
                ],
                declarations: [NamespaceSelectComponent],
                exports: [NamespaceSelectComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(NamespaceSelectModule, { declarations: [NamespaceSelectComponent], imports: [CommonModule,
        FormsModule,
        HttpClientModule,
        BrowserAnimationsModule,
        MatFormFieldModule,
        MatSelectModule,
        SnackBarModule], exports: [NamespaceSelectComponent] }); })();
