// kubeflow/esm2015/lib/date-time/date-time.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DateTimeComponent } from './date-time.component';
import { PopoverModule } from '../popover/popover.module';
import { ToDatePipe } from './to-date.pipe';
import { DetailsListModule } from '../details-list/details-list.module';
import * as i0 from "@angular/core";
export class DateTimeModule {
}
DateTimeModule.ɵfac = function DateTimeModule_Factory(t) { return new (t || DateTimeModule)(); };
DateTimeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: DateTimeModule });
DateTimeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, PopoverModule, DetailsListModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DateTimeModule, [{
        type: NgModule,
        args: [{
                declarations: [DateTimeComponent, ToDatePipe],
                imports: [CommonModule, PopoverModule, DetailsListModule],
                exports: [DateTimeComponent, ToDatePipe],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(DateTimeModule, { declarations: [DateTimeComponent, ToDatePipe], imports: [CommonModule, PopoverModule, DetailsListModule], exports: [DateTimeComponent, ToDatePipe] }); })();
