// kubeflow/esm2015/lib/date-time/to-date.pipe.js
import { Pipe } from '@angular/core';
import { defaultDateOptions, defaultTimeOptions, } from '../services/date-time.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/date-time.service";
export class ToDatePipe {
    constructor(dtService) {
        this.dtService = dtService;
    }
    transform(value, type) {
        if (!value) {
            return '';
        }
        try {
            const toDate = new Date(value);
            let options;
            if (type === 'date') {
                options = defaultDateOptions;
            }
            else if (type === 'time') {
                options = defaultTimeOptions;
            }
            return this.dtService.intlFormat(toDate, options);
        }
        catch (error) {
            console.error('ToDatePipe value:', value);
            console.error(error);
            return '';
        }
    }
}
ToDatePipe.ɵfac = function ToDatePipe_Factory(t) { return new (t || ToDatePipe)(i0.ɵɵdirectiveInject(i1.DateTimeService, 16)); };
ToDatePipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "libToDate", type: ToDatePipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ToDatePipe, [{
        type: Pipe,
        args: [{ name: 'libToDate' }]
    }], function () { return [{ type: i1.DateTimeService }]; }, null); })();
