// kubeflow/esm2015/lib/resource-table/status/status.component.js
import { Component, Input } from '@angular/core';
import { STATUS_TYPE } from './types';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/progress-spinner";
import * as i3 from "@angular/material/tooltip";
import * as i4 from "../../icon/icon.component";
import * as i5 from "@angular/material/icon";
function StatusComponent_mat_spinner_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "mat-spinner", 4);
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r0.config == null ? null : ctx_r0.config.getTooltip(ctx_r0.row));
} }
function StatusComponent_mat_spinner_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "mat-spinner", 4);
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r1.config == null ? null : ctx_r1.config.getTooltip(ctx_r1.row));
} }
function StatusComponent_lib_icon_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "lib-icon", 5);
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r2.config == null ? null : ctx_r2.config.getTooltip(ctx_r2.row));
} }
function StatusComponent_mat_icon_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "mat-icon", 6);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("matTooltip", ctx_r3.config == null ? null : ctx_r3.config.getTooltip(ctx_r3.row));
    i0.ɵɵproperty("ngClass", ctx_r3.config == null ? null : ctx_r3.config.getCssClasses(ctx_r3.row));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r3.config == null ? null : ctx_r3.config.getIcon(ctx_r3.row), " ");
} }
export class StatusComponent {
    constructor() {
        this.STATUS_TYPE = STATUS_TYPE;
    }
    ngOnInit() { }
}
StatusComponent.ɵfac = function StatusComponent_Factory(t) { return new (t || StatusComponent)(); };
StatusComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: StatusComponent, selectors: [["lib-status"]], inputs: { row: "row", config: "config" }, decls: 5, vars: 4, consts: [[3, "ngSwitch"], ["diameter", "24", "matTooltipClass", "custom-tooltip", 3, "matTooltip", 4, "ngSwitchCase"], ["icon", "custom:stoppedResource", 3, "matTooltip", 4, "ngSwitchCase"], ["matTooltipClass", "custom-tooltip", 3, "ngClass", "matTooltip", 4, "ngSwitchDefault"], ["diameter", "24", "matTooltipClass", "custom-tooltip", 3, "matTooltip"], ["icon", "custom:stoppedResource", 3, "matTooltip"], ["matTooltipClass", "custom-tooltip", 3, "ngClass", "matTooltip"]], template: function StatusComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementContainerStart(0, 0);
        i0.ɵɵtemplate(1, StatusComponent_mat_spinner_1_Template, 1, 1, "mat-spinner", 1);
        i0.ɵɵtemplate(2, StatusComponent_mat_spinner_2_Template, 1, 1, "mat-spinner", 1);
        i0.ɵɵtemplate(3, StatusComponent_lib_icon_3_Template, 1, 1, "lib-icon", 2);
        i0.ɵɵtemplate(4, StatusComponent_mat_icon_4_Template, 2, 3, "mat-icon", 3);
        i0.ɵɵelementContainerEnd();
    } if (rf & 2) {
        i0.ɵɵproperty("ngSwitch", ctx.config == null ? null : ctx.config.getPhase(ctx.row));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", ctx.STATUS_TYPE.WAITING);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", ctx.STATUS_TYPE.TERMINATING);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngSwitchCase", ctx.STATUS_TYPE.STOPPED);
    } }, directives: [i1.NgSwitch, i1.NgSwitchCase, i1.NgSwitchDefault, i2.MatSpinner, i3.MatTooltip, i4.IconComponent, i5.MatIcon, i1.NgClass], styles: [".status[_ngcontent-%COMP%]{display:flex;vertical-align:middle}.ready[_ngcontent-%COMP%]{color:green}.unavailable[_ngcontent-%COMP%]{color:grey}.warning[_ngcontent-%COMP%]{color:orange}.error[_ngcontent-%COMP%]{color:red}.stop[_ngcontent-%COMP%]{color:grey}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(StatusComponent, [{
        type: Component,
        args: [{
                selector: 'lib-status',
                templateUrl: './status.component.html',
                styleUrls: ['./status.component.scss'],
            }]
    }], function () { return []; }, { row: [{
            type: Input
        }], config: [{
            type: Input
        }] }); })();
