// kubeflow/esm2015/lib/confirm-dialog/dialog/dialog.component.js
import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA } from '@angular/material/dialog';
import { DIALOG_RESP } from '../types';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/dialog";
import * as i2 from "@angular/material/button";
import * as i3 from "@angular/common";
import * as i4 from "@angular/material/progress-spinner";
function ConfirmDialogComponent_button_10_Template(rf, ctx) { if (rf & 1) {
    const _r3 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 7);
    i0.ɵɵlistener("click", function ConfirmDialogComponent_button_10_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r3); const ctx_r2 = i0.ɵɵnextContext(); return ctx_r2.onAcceptClicked(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("color", ctx_r0.data.confirmColor);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r0.data.accept.toUpperCase(), " ");
} }
function ConfirmDialogComponent_button_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "button", 8);
    i0.ɵɵelementStart(1, "div", 9);
    i0.ɵɵelement(2, "mat-spinner", 10);
    i0.ɵɵelementStart(3, "p");
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(4);
    i0.ɵɵtextInterpolate(ctx_r1.data.applying);
} }
export class ConfirmDialogComponent {
    constructor(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.DIALOG_RESP = DIALOG_RESP;
        this.isApplying = false;
        this.applying$ = new Subject();
    }
    ngOnInit() {
        this.applying$.subscribe(b => {
            this.isApplying = b;
        });
    }
    onAcceptClicked() {
        this.isApplying = true;
        this.applying$.next(true);
    }
    onCancelClicked() {
        this.dialogRef.close(DIALOG_RESP.CANCEL);
    }
}
ConfirmDialogComponent.ɵfac = function ConfirmDialogComponent_Factory(t) { return new (t || ConfirmDialogComponent)(i0.ɵɵdirectiveInject(i1.MatDialogRef), i0.ɵɵdirectiveInject(MAT_DIALOG_DATA)); };
ConfirmDialogComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: ConfirmDialogComponent, selectors: [["lib-confirm-dialog"]], decls: 12, vars: 7, consts: [["mat-dialog-title", ""], ["mat-dialog-content", ""], [1, "error"], ["mat-dialog-actions", ""], ["mat-button", "", "cdkFocusInitial", "", 3, "mat-dialog-close"], ["mat-button", "", 3, "color", "click", 4, "ngIf"], ["mat-button", "", "disabled", "", 4, "ngIf"], ["mat-button", "", 3, "color", "click"], ["mat-button", "", "disabled", ""], [1, "waiting-button-wrapper"], ["diameter", "16"]], template: function ConfirmDialogComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "h1", 0);
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(2, "div", 1);
        i0.ɵɵelementStart(3, "p");
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(5, "p", 2);
        i0.ɵɵtext(6);
        i0.ɵɵelementEnd();
        i0.ɵɵelementEnd();
        i0.ɵɵelementStart(7, "div", 3);
        i0.ɵɵelementStart(8, "button", 4);
        i0.ɵɵtext(9);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(10, ConfirmDialogComponent_button_10_Template, 2, 2, "button", 5);
        i0.ɵɵtemplate(11, ConfirmDialogComponent_button_11_Template, 5, 1, "button", 6);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx.data.title);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.data.message);
        i0.ɵɵadvance(2);
        i0.ɵɵtextInterpolate(ctx.data.error);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("mat-dialog-close", ctx.DIALOG_RESP.CANCEL);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1(" ", ctx.data.cancel.toUpperCase(), " ");
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", !ctx.isApplying);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isApplying);
    } }, directives: [i1.MatDialogTitle, i1.MatDialogContent, i1.MatDialogActions, i2.MatButton, i1.MatDialogClose, i3.NgIf, i4.MatSpinner], styles: [".waiting-button-wrapper[_ngcontent-%COMP%]{display:flex}.waiting-button-wrapper[_ngcontent-%COMP%]   mat-spinner[_ngcontent-%COMP%]{margin:auto .2rem}.error[_ngcontent-%COMP%]{color:red}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ConfirmDialogComponent, [{
        type: Component,
        args: [{
                selector: 'lib-confirm-dialog',
                templateUrl: './dialog.component.html',
                styleUrls: ['./dialog.component.scss'],
            }]
    }], function () { return [{ type: i1.MatDialogRef }, { type: undefined, decorators: [{
                type: Inject,
                args: [MAT_DIALOG_DATA]
            }] }]; }, null); })();
