// kubeflow/esm2015/lib/loading-spinner/loading-spinner.component.js
import { Component, Input, ViewChild, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/material/progress-spinner";
const _c0 = ["spinnerWrapper"];
function LoadingSpinnerComponent_mat_spinner_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "mat-spinner", 3);
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("diameter", ctx_r1.diameter);
} }
const _c1 = function (a0) { return { height: a0 }; };
export class LoadingSpinnerComponent {
    constructor() {
        this.diameter = 32;
        this.height = `${this.diameter}px`;
        this.initialized = false;
    }
    ngAfterViewInit() {
        if (!this.wrapper) {
            return;
        }
        setTimeout(() => {
            const offset = this.wrapper.nativeElement.getBoundingClientRect().top;
            this.height = `calc(100vh - ${offset}px)`;
            this.initialized = true;
        });
    }
}
LoadingSpinnerComponent.ɵfac = function LoadingSpinnerComponent_Factory(t) { return new (t || LoadingSpinnerComponent)(); };
LoadingSpinnerComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: LoadingSpinnerComponent, selectors: [["lib-loading-spinner"]], viewQuery: function LoadingSpinnerComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.wrapper = _t.first);
    } }, inputs: { diameter: "diameter" }, decls: 3, vars: 4, consts: [[1, "spinner-wrapper", 3, "ngStyle"], ["spinnerWrapper", ""], [3, "diameter", 4, "ngIf"], [3, "diameter"]], template: function LoadingSpinnerComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0, 1);
        i0.ɵɵtemplate(2, LoadingSpinnerComponent_mat_spinner_2_Template, 1, 1, "mat-spinner", 2);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(2, _c1, ctx.height));
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.initialized);
    } }, directives: [i1.NgStyle, i1.NgIf, i2.MatSpinner], styles: [".spinner-wrapper[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:center;align-items:center}"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(LoadingSpinnerComponent, [{
        type: Component,
        args: [{
                selector: 'lib-loading-spinner',
                templateUrl: './loading-spinner.component.html',
                styleUrls: ['./loading-spinner.component.scss'],
            }]
    }], null, { diameter: [{
            type: Input
        }], wrapper: [{
            type: ViewChild,
            args: ['spinnerWrapper']
        }] }); })();
